### PR TITLE
Fix SR-4601 Report Added and Removed Benchmarks in Performance Comparison

### PR DESCRIPTION
**Pull request by palimondo**
Refactoring of `compare_perf_tests.py`.
Broken out of #8793.

Fix [SR-4601](https://bugs.swift.org/browse/SR-4601) Report Added and Removed Benchmarks in Performance Comparison
Fixes HTML partitions and sorting bugs introduced in #8923.

Doesn't yet work with @atrick's use case of merged multiple outputs from `Benchmark_Driver`. Will fix later...

### CODE REVIEW

**Comment by palimondo** [comment]
Please review @gottesmm @moiseev @atrick @shahmishal

**Comment by moiseev** [comment]
@swift-ci Please smoke benchmark

**Comment by palimondo** [comment]
@atrick Your use case, comparing concatenated `Benchmark_Driver` outputs you described in https://github.com/apple/swift/pull/8974#issuecomment-296848554, should work again.

**Comment by palimondo** [comment]
@moiseev @gottesmm @atrick I think this is ready for merge. Please review?

**Comment by moiseev** [comment]
Just tried to run the refactored script on my benchmark result files that are clearly different, and it produced nothing.

**Comment by palimondo** [comment]
> On 26 Apr 2017, at 00:32, Maxim Moiseev <notifications@github.com> wrote:
> 
> Just tried to run the refactored script on my benchmark result files that are clearly different, and it produced nothing.
> 
Hmm.. Could you send me your logs to test?

File: benchmark/scripts/compare_perf_tests.py
Comment (moiseev) [inline benchmark/scripts/compare_perf_tests.py]: Benchmark output contains exactly 8 columns, so all of them get skipped because of this line, I believe.

File: benchmark/scripts/compare_perf_tests.py
Comment (moiseev) [inline benchmark/scripts/compare_perf_tests.py]: This line also breaks when I ask this script to compare benchmark executable output, as there are only 8 columns there.

**Comment by moiseev** [comment]
Oh, it’s just the output of Benchmark_O, that I feed into the script with —old-file and —new-file args.

**Comment by palimondo** [comment]
On Wed, 26 Apr 2017 at 00:56, Maxim Moiseev <notifications@github.com>
wrote:

> Oh, it’s just the output of Benchmark_O, that I feed into the script with
> —old-file and —new-file args.


That explains it. I built it with expectation of processing
`Benchmark_Driver`'s output that has MAX_RSS column as last, for total of 9.

I see you are all fans of manual work, (@atrick ;-) ), I'll get back to
you. Initially I went with the simplest condition that worked with the way
automated scripts use `compare_perf_tests.py`, but I'm slowly discovering
the homegrown forest of per-person customized use-cases that are not
visible from the outside. I'll add documentation, that will make this
clearer for the future, too.

>

**Comment by palimondo** [comment]
We should be good to land this now?

**Comment by moiseev** [comment]
FWIW: ran it locally with my set of benchmark results and it works as expected (including the changes that I tried to make).
I'm still not a big fan of the `TestComparator`, class whose only functionality is in the constructor, but it's OK for the script.
:shipit: from me.
@shahmishal, it's your call.
Thanks @palimondo!

**Comment by moiseev** [comment]
@swift-ci Please smoke benchmark

**Review by shahmishal — APPROVED** [review]
LGTM! Lets wait for the benchmark run to complete before merging this PR. Thanks @palimondo and @moiseev!

**Comment by shahmishal** [comment]
@swift-ci Please smoke test

**Comment by shahmishal** [comment]
Waiting for this job to complete before merging this PR -> https://ci.swift.org/job/swift-PR-osx-smoke-perf/38/

**Pull request merged**
